libcasr/src/sarif.rs

```diff
@@ -0,0 +1,404 @@
+        );
+        let mut full_desc = Map::new();
+        full_desc.insert("text".to_string(), Value::String(class.explanation.clone()));
+        rule.insert(
+            "fullDescription".to_string(),
+            Value::Object(full_desc.clone()),
+        );
+        let mut properties = Map::new();
+        properties.insert(
+            "problem.severity".to_string(),
+            Value::String(class.severity.clone()),
+        );
+        rule.insert("properties".to_string(), Value::Object(properties));
+
+        (Some(Value::Object(rule)), rule_id)
+    }
+
+    /// Add CASR CrashReport to SARIF report.
+    /// NOTE: before using this method,
+    /// use SarifReprot::new() to get report.
+    ///
+    ///  # Arguments
+    ///
+    ///  * 'report' - CrashReport
+    ///
+    ///  * 'source_root' - Path to source root directory.
+    pub fn add_casr_report<T: AsRef<Path>>(
+        &mut self,
+        report: &CrashReport,
+        source_root: T,
+    ) -> Result<()> {
+        let (rule, rule_id) = self.rule(&report.execution_class);
+        if let Some(rule) = rule {
+            let rules = self.json.as_object_mut().unwrap()["runs"]
+                .as_array_mut()
+                .unwrap()[0]
+                .as_object_mut()
+                .unwrap()["tool"]
+                .as_object_mut()
+                .unwrap()["driver"]
+                .as_object_mut()
+                .unwrap()["rules"]
+                .as_array_mut()
+                .unwrap();
+            rules.push(rule);
+            if rule_id.starts_with('G') {
+                self.id += 1;
+            }
+        }
+
+        let results = self.json.as_object_mut().unwrap()["runs"]
+            .as_array_mut()
+            .unwrap()[0]
+            .as_object_mut()
+            .unwrap()["results"]
+            .as_array_mut()
+            .unwrap();
+        let mut result = Map::new();
+        result.insert("ruleId".to_string(), Value::String(rule_id));
+        result.insert("level".to_string(), Value::String("error".to_string()));
+        let mut message = Map::new();
+        let text = if !report.stdin.is_empty() {
+            format!(
+                "{}: {} < {}",
+                report.execution_class.short_description, report.proc_cmdline, report.stdin
+            )
+        } else {
+            format!(
+                "{}: {}",
+                report.execution_class.short_description, report.proc_cmdline
+            )
+        };
+        message.insert("text".to_string(), Value::String(text));
+        result.insert("message".to_string(), Value::Object(message));
+        let mut locations: Vec<Value> = Vec::new();
+        let mut location = Map::new();
+        let mut physical_loc = Map::new();
+        let mut artifact_loc = Map::new();
+        let mut region = Map::new();
+        let parts: Vec<_> = report.crashline.split(':').map(|s| s.to_string()).collect();
+        if parts.len() != 2 && parts.len() != 3 {
```

Comment:
crash line may miss a line number, this is a real case

Reply:
https://github.com/ispras/casr/blob/master/casr/src/bin/casr-dojo.rs#L443

Reply:
Okay, I understand but what we want to do in this case. I think just return an error.